query/services/ingestion.py

```diff
@@ -0,0 +1,199 @@
+import logging
+import math
+from datetime import datetime, timezone
+from typing import Dict
+
+from asyncpg import Connection
+
+from query.database import database_connection
+from query.models.product import Source
+from query.mongodb import find_products
+from query.tables.loaded_tag import append_loaded_tags
+from query.tables.product import (
+    create_minimal_product,
+    delete_products,
+    get_minimal_product,
+    product_fields,
+    update_products_from_staging,
+)
+from query.tables.product_country import fixup_product_countries
+from query.tables.product_ingredient import create_ingredients_from_staging
+from query.tables.product_tags import TAG_TABLES, create_tags_from_staging
+from query.tables.settings import get_last_updated, set_last_updated
+
+logger = logging.getLogger(__name__)
+min_datetime = datetime(1, 1, 1, tzinfo=timezone.utc)
+batch_size = 10000
+
+
+async def get_process_id(connection):
+    return await connection.fetchval("SELECT pg_current_xact_id()")
+
+
+def int_or_none(value):
+    return None if value == None else int(value)
+
+
+async def import_with_filter(filter: Dict, source: Source) -> datetime:
+    max_last_updated = min_datetime
+    found_product_codes = []
+    codes_specified = "code" in filter and "$in" in filter["code"]
+
+    async with database_connection() as connection:
+        await connection.execute(
+            "CREATE TEMP TABLE product_temp (id int PRIMARY KEY, last_updated timestamptz, data jsonb)"
+        )
+        try:
+            process_id = await get_process_id(connection)
+            projection = {
+                key: True
+                for key in (list(TAG_TABLES.keys()) + list(product_fields.keys()))
```

Comment:
You use list(TAG_TABLES.keys()) + list(product_fields.keys()) in different places, I'm wondering if it should not be a small function (to avoid missing some part of the code, if this changes)